pkg/raft/log.go

```diff
@@ -531,6 +550,27 @@ func (l *raftLog) scan(lo, hi uint64, pageSize entryEncodingSize, v func([]pb.En
 // The returned slice can be appended to, but the entries in it must not be
 // mutated.
 func (l *raftLog) slice(lo, hi uint64, maxSize entryEncodingSize) ([]pb.Entry, error) {
+	return l.snap().slice(lo, hi, maxSize)
+}
+
+// LogSlice returns a valid log slice for the (lo, hi] log index interval.
+func (l LogSnapshot) LogSlice(lo, hi uint64, maxSize uint64) logSlice {
+	prevTerm, err := l.term(lo)
+	if err != nil { // the log is probably compacted at index > lo
+		return logSlice{}
```

Comment:
Shouldn't this method be able to return an error? After all, it's doing IO

Reply:
Done. The storage error handling is a bit all over the place. Some methods panic, some panic only if error is not `ErrCompacted`, etc. There is a TODO to unify all that.